cpp/src/arrow/memory_pool.cc

```diff
@@ -182,6 +182,12 @@ class SystemAllocator {
     if (!*out) {
       return Status::OutOfMemory("malloc of size ", size, " failed");
     }
+#elif defined(sun) || defined(__sun)
+#include <stdlib.h>
+    *out = reinterpret_cast<uint8_t*>(memalign(static_cast<size_t>(size), kAlignment));
```

Comment:
The [signature](https://docs.oracle.com/cd/E36784_01/html/E36874/memalign-3c.html) is `void *memalign(size_t alignment, size_t size)`, so you've got the arguments flipped (did you run the tests?).

Reply:
Good catch, thanks. I do not have a Solaris dev environment, but there is a service I can use to test R package building on Solaris, and this fixed the previous compile error on it. Because of all sorts of other issues, I haven't managed to get a full successful build yet (I'm up to 50% at the latest), so I haven't gotten to the point of seeing if the code actually works.